Doc/library/sqlite3.rst

```diff
@@ -1434,10 +1434,13 @@ case-insensitively by name:
 Using the connection as a context manager
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-Connection objects can be used as context managers
-that automatically commit or rollback transactions.  In the event of an
```

Comment:
"... that automatically commit or rollback transactions that occur within the context." would probably be the most precise statement of this (if I understand it correctly).

----

That said, further clarification question: if there is an uncommitted transaction `t1` on `con` prior to entering the context (as in the example in the docs), what happens to it at `__exit__`? Based on my understanding and the various versions of the text, I would expect:

- If *no transactions* are made inside the context, then `t_1` remains on `con`, but in an uncommitted state (neither committed nor rolled back -- the context is a no-op)
- If one or more transactions `t_j` is added to `con` within the context *AND* if no uncaught error occurs in the context, I would expect `t1` to be committed along with all the `t_j` at context exit
- If one or more transactions `t_j` is added to `con` within the context *AND* an uncaught exception occurs, I would expect `t1` to be rolled back along with all of the `t_j` at context exit.

Regardless of whether the above is correct, I think some mention needs to be made about the disposition of pre-existing transactions on `con`.

Reply:
If what actually happens in the following example is that `t_1` is committed at `__exit__`, then the docs *definitely* need to call that out:

```
con = sqlite3.connect(...)
con.execute(sql_1)

with con:
    pass

# Is sql_1 now committed at this point?
```

Reply:
I'll try to make this explicit, technically:

- `__enter__`: nothing happens; returns self
- `__exit__`: if a transaction is open, commit or rollback; if commit failed, roll back; if no transaction is open, nothing happens

Now, we need to write this in a userfriendly way, trying to put emphasis on "the happy path".

Reply:
Does this fit the semantics?
```rest
Connection objects can be used as context managers that automatically commit or
rollback open transactions when leaving the body of the context manager.
If the body of the :keyword:`with` statement finishes without exceptions,
the transaction is committed.
If an exception is raised and not caught, the transaction is rolled back.

If there is no open transaction on leaving the body of the :keyword:`with`
statement, the context manager is a no-op.
```

A

Reply:
Tha ks for your suggestions all. I think we should avoid being too verbose. Expect an update tonight (CET).

Reply:
> The exit processing is to call commit (unless there was an uncaught exception), and that's already well-defined [...]

But that is not the reality (!), and we cannot change the status quo without breaking people's code. What we _can_ do, however, is document the actual behaviour, to get rid of the confusion, so that is what we're trying to do here :)

I'll repeat what I said earlier in this thread, a little bit more verbose:

`__enter__` does nothing ... the current confusion starts already here, because a lot of people _think_ that `__enter__` opens a new transaction! But it does not, and assuming otherwise can bite your code, and you can end up debugging your code for days without finding out what is wrong. Take a look at the linked issue, and you'll see the complaints ;) This is why _I_ think it is a good idea to document, and maybe even add emphasis to, the fact that _no transaction is opened when you enter the context manager_. The original commit of this PR had a `note` directive for this fact.

`__exit__`; this is the source of the rest of the confusion. A lot of people _think_ that the context manager will commit/rollback at exit. But that is not the case. Commit/rollback is only performed _if there is an open transaction_. Since the context manager (confusingly) does _not_ open a transaction, but people assume so, we get complains, because people experience unexpected behaviour. In the included example in the docs, there is an implicit transaction opened _because a DML query is executed during the with body_, so at `__exit__`, we can expect commit or rollback.

As you'll see from the issue, the original pysqlite author is +1 to deprecating the context manager, and I fully understand that :)

Reply:
All right, PTAL! :)

Reply:
And to add to the confusion: if `isolation_level` is not `None` and you are executing a DML statement, a transaction is implicitly started, for `execute` and `executemany`, but not for `executescript`.

Reply:
> > The exit processing is to call commit (unless there was an uncaught exception), and that's already well-defined [...]
> 
> But that is not the reality (!), and we cannot change the status quo without breaking people's code. 

I assume I'm missing something (outer wrapping around this maybe?), but this is what I see in the sqlite module code, which looks that *is* what it does:

```C
static PyObject *
pysqlite_connection_exit_impl(pysqlite_Connection *self, PyObject *exc_type,
                              PyObject *exc_value, PyObject *exc_tb)
/*[clinic end generated code: output=0705200e9321202a input=bd66f1532c9c54a7]*/
{
    int commit = 0;
    PyObject* result;

    if (exc_type == Py_None && exc_value == Py_None && exc_tb == Py_None) {
        commit = 1;
        result = pysqlite_connection_commit_impl(self);
    }
    else {
        result = pysqlite_connection_rollback_impl(self);
    }
    ...
```

Reply:
You _assume_ what those functions are doing. Now, follow those paths one level further.

Reply:
Even better, you can test the functionality using the sqlite3 extension module. The connection attribute `in_transaction` is mapped _directly_ to the _inverted_ result of `sqlite3_get_autocommit`.

<details>
<summary>Test: open transaction before ctx mgr</summary>

```py
>>> import sqlite3
>>> cx = sqlite3.connect(":memory:")
>>> cx.set_trace_callback(print)
>>> cx.in_transaction
False
>>> cx.execute("BEGIN")  # explicitly opening a transaction
BEGIN
<sqlite3.Cursor object at 0x7fd85413aa40>
>>> cx.in_transaction
True
>>> with cx: pass
...
COMMIT
>>> cx.in_transaction
False
```

</details>

<details>
<summary>Test: no transaction before ctx mgr</summary>

```py
>>> import sqlite3
>>> cx = sqlite3.connect(":memory:")
>>> cx.set_trace_callback(print)
>>> cx.in_transaction
False
>>> with cx: pass
...
>>> cx.in_transaction
False
```

</details>

<details>
<summary>Test: implicit transaction opened during ctx mgr</summary>

```py
>>> import sqlite3
>>> cx = sqlite3.connect(":memory:")
>>> cx.set_trace_callback(print)
>>> cx.execute("create table t(t)")
create table t(t)
<sqlite3.Cursor object at 0x7f4a06d72a40>
>>> cx.in_transaction
False
>>> with cx: cx.execute("insert into t values(1)")
...
BEGIN
insert into t values(1)
<sqlite3.Cursor object at 0x7f4a06d72a40>
COMMIT
>>> cx.in_transaction
False
```

</details>

Reply:
I consider this conversation resolved. If there are more questions, please start a new conversation.

Reply:
> You _assume_ what those functions are doing. Now, follow those function calls one level further.

So (real question, no criticism): does that mean there ought to be more detail in the commit/rollback descriptions, rather than the context manager bit?